### PR TITLE
Fix hide/close behaviour on Windows, Linux and MacOS

### PR DESCRIPTION
**Pull request by cookiengineer**
Before submitting, please confirm you've
 - [x] read and understood our [Contributing Guidelines](https://github.com/mattermost/desktop/blob/master/CONTRIBUTING.md)
 - [x] completed [Mattermost Contributor Agreement](http://www.mattermost.org/mattermost-contributor-agreement/)
 - [x] executed `npm run lint:js` for proper code formatting

Please provide the following information:

**Summary**

This pull request fixes issue #90 on Windows 7, Windows 10, MacOS Mojave (don't have Catalina to test), and Arch Linux (GNOME + KDE both tested).

- [x] The commit respects `config.minimizeToTray`
- [x] Correct and user-expected behaviour that closes window when quitting the application, if `minimizeToTray` is `false`.

**Issue link**

https://github.com/mattermost/desktop/issues/90

### CODE REVIEW

File: src/main/mainWindow.js [hunk 1]
```diff
@@ -131,15 +131,22 @@ function createMainWindow(config, options) {
         window.blur(); // To move focus to the next top-level window in Windows
         window.hide();
       }
+      function closeWindow(window) {
```
[the remark this answers is not in the document]
Comment (cookiengineer) [reply src/main/mainWindow.js:134]: I was assuming that this is preferred, hence the hideWindow inner function that basically has only two lines. Should this be disregarded?

Then I would just call the methods on mainWindow directly within the switch/case block.

File: src/main/mainWindow.js [hunk 2]
```diff
@@ -131,15 +131,22 @@ function createMainWindow(config, options) {
...
+        window.close();
+      }
       switch (process.platform) {
       case 'win32':
-        hideWindow(mainWindow);
+        if (config.minimizeToTray) {
```
[the remark this answers is not in the document]
Comment (cookiengineer) [reply src/main/mainWindow.js:139]: I think it is integrated in the Settings page: https://github.com/mattermost/desktop/blob/9b5f53a294f05c4b2ad41c7d220e1e7f4064ee02/src/browser/components/SettingsPage.jsx#L307

Why there was such a decision, and whether or not the UX is intended to be the same on all platforms, I cannot decide.

File: src/main/mainWindow.js [hunk 3]
```diff
@@ -149,8 +156,10 @@ function createMainWindow(config, options) {
             app.hide();
           });
           mainWindow.setFullScreen(false);
-        } else {
+        } else if (config.minimizeToTray) {
```
[the remark this answers is not in the document]
Comment (cookiengineer) [reply src/main/mainWindow.js:159]: I think it is integrated in the Settings page: https://github.com/mattermost/desktop/blob/9b5f53a294f05c4b2ad41c7d220e1e7f4064ee02/src/browser/components/SettingsPage.jsx#L307

Why there was such a decision, and whether or not the UX is intended to be the same on all platforms, I cannot decide.

[the remark this answers is not in the document]
Comment (Willyfrog) [reply src/main/mainWindow.js:159]: seems like for osX is always false, what I meant is that you don't have the option in the settings page for turning it on.  And since there is no option for windows as well, I think it'll show as `undefined` although i haven't check, there might be a default option from the config as well. What I mean is that on win and osx there is no choice for the user.

File: src/main/mainWindow.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (Willyfrog) [reply src/main/mainWindow.js:134]: I guess it was for adding that blur in front, but I still think there is not much to gain from that other than giving it a name.

File: src/main/mainWindow.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (cookiengineer) [reply src/main/mainWindow.js:159]: So, would it be possible to simply add this setting to the other platforms, too?

What was the reason to not add this to MacOS or Windows in the first place?

https://github.com/mattermost/desktop/blob/9b5f53a294f05c4b2ad41c7d220e1e7f4064ee02/src/browser/components/SettingsPage.jsx#L847

^ The option should be easy to integrate with MacOS and Windows...

[the remark this answers is not in the document]
Comment (Willyfrog) [reply src/main/mainWindow.js:159]: @esethna maybe you can comment on this

[the remark this answers is not in the document]
Comment (esethna) [reply src/main/mainWindow.js:159]: @Willyfrog can you summarize what the user experience change is and what's expected? I'm not sure what's being proposed here. Please ask in the desktop app contributors channel if this is a separate proposed feature enhancement

[the remark this answers is not in the document]
Comment (Willyfrog) [reply src/main/mainWindow.js:159]: created conversation on [comminity server](https://community-daily.mattermost.com/core/pl/sndhqk8ga7nu9g33ru96np74uy)

[the remark this answers is not in the document]
Comment (Willyfrog) [reply src/main/mainWindow.js:159]: as per the above conversation, it would be better to restrict the bugfix to linux and file a separate PR if we want to have consistency (I'm 0/5 on this as long as it's optional)

[the remark this answers is not in the document]
Comment (cookiengineer) [reply src/main/mainWindow.js:159]: @Willyfrog So that means the decision was commonly agreed to move against the UX consistency and do not want to have this integrated with MacOS and Windows (as per removal of the platform check in the Settings jsx file for those platforms)?

[the remark this answers is not in the document]
Comment (Willyfrog) [reply src/main/mainWindow.js:159]: For this PR, yes. It would be preferred.
But that doesn't mean we don't want to have the option, but it would be a new feature, so it would be better if we do it separately

[the remark this answers is not in the document]
Comment (cookiengineer) [reply src/main/mainWindow.js:159]: @Willyfrog Who is the former code owner of that decision? Why was it built for Linux only in the first place? What are the pitfalls I would have to test against?

[the remark this answers is not in the document]
Comment (Willyfrog) [reply src/main/mainWindow.js:159]: Since I wasn't around when this happen, I can only guess that it was because windows and mac default behaviour is different from linux.
On Win, if there is a tray icon it is expected to close the window but have the program running. if there isn't one, then it is expected to quit the app.
On mac, it is expected to keep running wether there is a tray icon or not, and quit on `cmd+q`.

regarding pitfalls, first thing that comes to mind is configuration upgrade and making sure that the app blurs when it is closed/hidden.

[the remark this answers is not in the document]
Comment (cookiengineer) [reply src/main/mainWindow.js:159]: Regarding Windows: This is the identical behaviour as it is expected on Linux (both GNOME/MATE and KDE).

Regarding MacOS:

- Close Window button must close a window, see this: https://support.apple.com/en-gb/guide/mac-help/mchlp2469/mac
- Restore view feature is missing in general, and Mattermost only has one window; so this is kinda does not apply to Mattermost: https://support.apple.com/en-gb/guide/mac-help/mchl834d18c2/10.15/mac/10.15

[the remark this answers is not in the document]
Comment (Willyfrog) [reply src/main/mainWindow.js:159]: yes, i think it is working that way, right?

The new feature would be to allow the user to choose it. Like I said, it's not that we don't want it, but have a separate PR so we make sure the bug can be closed wether the feature makes it or not.

**Comment by jasonblais** [comment]
Hi @cookiengineer - thanks for your contribution! Let us know if we can help with changes or if you have any questions about the feedback above. We'd be happy to help :)

**Comment by cookiengineer** [comment]
@jasonblais Currently I'm unsure how to proceed, as this PR will not be merged until all UI/UX relevant to that "minimize to tray" setting is created, approved, and implemented.

So, a mattermost core member responsible for that has to create the designs (if necessary), and help me regarding this; as I have no clue how to proceed.

Currently I'm basically maintaining my own fork for my company, because pretty much every user (from Linux, Windows and MacOS Catalina+) I've encountered so far was annoyed by the default behaviour.

**Comment by jasonblais** [comment]
@cookiengineer Sorry about it, I had missed that part of the conversation.

It looks like the proposal was to apply the fix for Linux only - would that be an option?

**Comment by Willyfrog** [comment]
hello @cookiengineer 

let me know if this works for you:
1) limiting this PR to just the linux change
2) talk in the desktop channel (we can continue on [this thread if you like](https://community-daily.mattermost.com/core/pl/sndhqk8ga7nu9g33ru96np74uy) ) to involve PM/UX and see what's the best approach for having consistency on all platforms and expected behaviour.

**Comment by Willyfrog** [comment]
closing, as codebase has changed a lot and we would probably need to see if it still applies.

Happy to reopen if there is a chance to reuse it.

**Comment by cookiengineer** [comment]
Nobody helped out from the PM/UX side here, so I never assumed this is gonna be merged anyways :man_shrugging: As long as politics stand in the way of approving a simple PR as this one, I think the workflow is overoptimized - well, or aims to be.

But this is probably a discussion for someone else, I stopped using mattermost because of this.

**Comment by Willyfrog** [comment]
I'm sorry you felt this way, and yes, sometimes we need to have some process to ensure everything goes as smoothly as we can for everyone involved, that doesn't mean the process is perfect, but we try to make it as easy as we can for all.

Let me or @jasonblais  know if there is any specific part of the process that we should improve, so we can take a look at it with perhaps a new point of view.

Thank you for pouring some of you time into this PR, that is always appreciated.
